### PR TITLE
Add timeout to engine object

### PR DESCRIPTION
**Pull request by dstrain115**
- Adds timeout to Engine object and makes it configurable
- This plumbs the timeout down to results object so that the
amount of polling is configurable.

Solves #2817

### CODE REVIEW

File: cirq/google/engine/engine.py [hunk 1]
```diff
@@ -138,6 +139,8 @@ def __init__(
                 configure options on the underlying client.
             verbose: Suppresses stderr messages when set to False. Default is
                 true.
+            timeout: Timeout for polling for results, in seconds.  Default is
+                to never timeout
```
Comment (dabacon) [inline cirq/google/engine/engine.py:143]: nit on period at end of comments.

File: cirq/google/engine/engine.py [hunk 2]
```diff
@@ -150,6 +153,7 @@ def __init__(
                                     service_args=service_args,
                                     verbose=verbose)
         self.context = context
```
Comment (dabacon) [inline cirq/google/engine/engine.py:155]: does it make sense to put this on the context?

File: cirq/google/engine/engine.py (hunk 1 above)
Comment (dstrain115) [reply cirq/google/engine/engine.py]: Done.

File: cirq/google/engine/engine.py (hunk 2 above)
Comment (dstrain115) [reply cirq/google/engine/engine.py:160]: Sounds like a good idea to me.  Done.

File: cirq/google/engine/engine_job.py [hunk 3]
```diff
@@ -262,10 +262,15 @@ def results(self) -> List[study.TrialResult]:
...
+            total_seconds_waited = 0.0
+            timeout = self.context.timeout
+            while True:
+                if timeout and total_seconds_waited >= timeout:
+                    break
                 if job.execution_status.state in TERMINAL_STATES:
                     break
                 time.sleep(0.5)
+                total_seconds_waited += 0.5
                 job = self._refresh_job()
```
Comment (dabacon) [inline cirq/google/engine/engine_job.py:274]: I guess this will not be exactly the timeout time since refresh job takes time.  But probably ok.

**Pull request merged**